src/test/java/org/springframework/data/r2dbc/testing/PostgresTestSupport.java

```diff
@@ -31,30 +35,60 @@
 	public static String INSERT_INTO_LEGOSET = "INSERT INTO legoset (id, name, manual) VALUES($1, $2, $3)";
 
 	/**
-	 * Returns a locally provided database at {@code postgres:@localhost:5432/postgres}.
+	 * Returns a database either hosted locally at {@code postgres:@localhost:5432/postgres} or running inside Docker.
 	 *
-	 * @return
+	 * @return information about the database. Guaranteed to be not {@literal null}.
 	 */
 	public static ExternalDatabase database() {
-		return local();
+
+		ExternalDatabase local = local();
+		if (local.checkValidity()) {
```

Comment:
I would suggest inverting the condition. "If Docker is available, otherwise".
Some users might have an important state in their local database, or an outdated version of it, or many other reasons.

FYI you can check Docker's availability with:
```java
return DockerClientFactory.instance().client() != null
```

Reply:
I understand and somewhat agree with the concern. 
My idea behind making it in this order was that one might want to use a specially configured database, e.g. for reproducing an issue that only appears with a special configuration.

I'm not at all sure if that is a relevant use case but we would lose it if we reverse the preference.

Reply:
I see. What about something like `Boolean.getProperty("spring.data.r2dbc.test.forceLocalDB")`?